rfcs/libraries.md

```diff
@@ -0,0 +1,128 @@
+local MyModule = require("../MyModule")
+ 
+-- From C:/MyOtherLibrary/MainModule.luau
+local MyModule = require("../MyLibrary/MyModule")
+```
+
+All relative paths will start with either `./`  or `../`  which denote the directory of the script or file, and parent directory respectively.
+
+#### Aliases
+
+Aliases can be used to bind an absolute or relative path to a convenient name that can be required directly. They are always prefixed with `$` to make it obvious they are not the same as a regular path.
+
+Each library has its own map which will be stored in the `.luaurc` file. Since libraries can be embedded inside of one another, any aliases which are not overriden will be inherited from the parent library.
+
+```json
+"Aliases": {
+    "Roact": "C:/LuauModules/Roact-v1.4.2"
+}
+```
+
+Based on that map you would be able to require Roact directly:
+
+```lua
+local Roact = require("$Roact")
+```
+
+Or even a sub-module:
+
+```lua
+local createElement = require("$Roact/createElement")
+```
+
+##### Versioning
+
+Aliases are simple bindings and aren't concerned with versioning. The intention is for a package manager to leverage aliases by automatically adding and updating the alias map to reflect a packages dependencies.
+
+##### Root Alias
+
+All libraries have a special alias for referring to their root. This alias cannot be overriden and is simply defined as `$`.
+
+Requiring any file in the library can be done relative to the root:
+```lua
+local ModuleAtLibraryRoot = require("$/ModuleAtLibraryRoot")
+```
+
+##### Limitations
+
+- Aliases cannot reference other aliases
+- Aliases cannot contain certain characters such as `/` or `$` (full list TBD)
+- All aliases must be prefixed with $ when used in a require statement to avoid ambiguity
+- Aliases can only occur at the beginning of a path
+- Multiple aliases are not supported
+
+#### Directories
+
+If the string resolves to a directory rather than a file then we will attempt to require a specific file in that directory with the following name:
+1. `init.lua`
+2. `init.luau`
+
+#### DataModel as VFS
+
+In the Roblox engine, the DataModel will act as a virtual file system. At the root of this VFS is the DataModel itself. This will allow packages from Luau to be imported and exported freely without needing to consider the platform they are being used on.
+
+All paths used in the Roblox engine must refer to a location in the DataModel, they cannot be used to access files on disk.
+
+#### Platforms
+
+For compatability across platforms, we will automatically map `/` onto `\` on Windows.
+
+#### Backwards Compatibility
+
+Luau libraries are not compatible with existing Lua libraries. This is because Lua favors the `.` based require syntax instead.
+
+- Libraries are fully compatible with the Roblox engine as string-syntax is unsupported.
+- Compatibility with existing Luau is TBD.
+
+### Defining a Library
+
+Any directory containing a `.luaurc` file is considered a library.
```

Comment:
A single library may contain multiple `.luaurc` files at various levels of directory tree, suppressing/tweaking linter settings for some files.
Mapping each `.luaurc` to `Library` unconditionally would also reset root alias `$` to these subfolders.
This can be unexpected to library authors.

Potential solution:
- use a dedicated key which once met defines current root alias. Its propagated down the tree and can be overriden
- use different file to define Library root

Reply:
This is a valid point. I'm unclear how we would import such a library though since we don't have `.luaurc` files in the game-engine. Will think about this some more and get back to you.

Reply:
I've given this some more thought and my proposed solution is to move the alias map out of `.luaurc` and into a generic `luauconfig.json` file instead.

This separates linting rules from library definition and creates a file for other configuration information that we might want to include in the future.